src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/CompletionSource.cs

```diff
@@ -234,15 +226,132 @@ static void CheckForExperimentStatus(ITextView textView, in CompletionOptions op
+            // contemplate such action thus typing slower before commit and/or spending more time examining the list, which give us some opportunities
+            // to still provide those items later before they are truly required.     
+
+            var options = _globalOptions.GetCompletionOptions(document.Project.Language);
+
+            if (!options.ShouldShowItemsFromUnimportNamspaces())
+            {
+                // No need to trigger expanded providers at all if the feature is disabled, just trigger core providers and return;
+                var (context, list) = await GetCompletionContextWorkerAsync(document, trigger, triggerLocation,
+                    options with { ExpandedCompletionBehavior = ExpandedCompletionMode.NonExpandedItemsOnly }, cancellationToken).ConfigureAwait(false);
+
+                AddPropertiesToSession(session, list, triggerLocation);
+                return context;
+            }
+            else if (!session.TextView.Options.GetOptionValue(DefaultOptions.ResponsiveCompletionOptionId))
+            {
+                // We tie the behavior of delaying expand items to editor's "responsive completion" option.
+                // i.e. "responsive completion" disabled == always wait for all items to be calculated.
+                var (context, list) = await GetCompletionContextWorkerAsync(document, trigger, triggerLocation,
+                    options with { ExpandedCompletionBehavior = ExpandedCompletionMode.AllItems }, cancellationToken).ConfigureAwait(false);
+
+                AddPropertiesToSession(session, list, triggerLocation);
+                AsyncCompletionLogger.LogImportCompletionGetContext(isBlocking: true, delayed: false);
+                return context;
+            }
+            else
             {
-                return AsyncCompletionData.CompletionContext.Empty;
+                // OK, expand item is enabled but we shouldn't block completion on its results.
+                // Kick off expand item calculation first in background.
+                Stopwatch stopwatch = new();
+                var expandedItemsTask = Task.Run(async () =>
+                {
+                    var result = await GetCompletionContextWorkerAsync(document, trigger, triggerLocation,
+                      options with { ExpandedCompletionBehavior = ExpandedCompletionMode.ExpandedItemsOnly }, cancellationToken).ConfigureAwait(false);
+
+                    // Record how long it takes for the background task to complete *after* core providers returned.
+                    // If telemetry shows that a short wait is all it takes for ExpandedItemsTask to complete in
+                    // majority of the sessions, then we might consider doing that instead of return immediately.
+                    var elapsedMs = (int)stopwatch.ElapsedMilliseconds;
+                    if (elapsedMs > 0)
```

Comment:
not sure why you have the > 0 check.

Reply:
This code is using the high-resolution timer, so the `>0` check is unnecessary.